src/puffin/src/partial_reader/async.rs

```diff
@@ -13,184 +13,122 @@
+    async fn metadata(&mut self) -> io::Result<Metadata> {
+        Ok(Metadata {
+            content_length: self.size,
+        })
+    }
 
-        // end of portion
-        if self.is_eof() {
-            return Poll::Ready(Ok(0));
+    async fn read(&mut self, range: Range<u64>) -> io::Result<Bytes> {
+        let absolute_range_start = self.offset + range.start;
+        if absolute_range_start >= self.offset + self.size {
+            return Err(io::Error::new(
+                io::ErrorKind::UnexpectedEof,
+                "Start of range is out of bounds",
+            ));
         }
+        let absolute_range_end = (self.offset + range.end).min(self.offset + self.size);
+        let absolute_range = absolute_range_start..absolute_range_end;
 
-        // first read, seek to the correct offset
-        if self.position_in_portion.is_none() {
-            // seek operation
-            let seek_from = io::SeekFrom::Start(self.offset);
-            ready!(self.as_mut().project().source.poll_seek(cx, seek_from))?;
+        let result = self.source.read(absolute_range.clone()).await?;
+        self.position_in_portion = Some(absolute_range.end);
+        Ok(result)
+    }
 
-            self.position_in_portion = Some(0);
+    async fn read_into(
+        &mut self,
+        range: Range<u64>,
+        buf: &mut (impl BufMut + Send),
+    ) -> io::Result<()> {
+        let absolute_range_start = self.offset + range.start;
+        if absolute_range_start >= self.offset + self.size {
+            return Err(io::Error::new(
+                io::ErrorKind::UnexpectedEof,
+                "Start of range is out of bounds",
+            ));
         }
+        let absolute_range_end = (self.offset + range.end).min(self.offset + self.size);
+        let absolute_range = absolute_range_start..absolute_range_end;
 
-        // prevent reading over the end
-        let max_len = (self.size() - self.position_in_portion.unwrap()) as usize;
-        let actual_len = max_len.min(buf.len());
-
-        // create a limited reader
-        let target_buf = &mut buf[..actual_len];
+        self.source.read_into(absolute_range.clone(), buf).await?;
+        self.position_in_portion = Some(absolute_range.end);
+        Ok(())
+    }
 
-        // read operation
-        let read_bytes = ready!(self.as_mut().project().source.poll_read(cx, target_buf))?;
-        self.position_in_portion = Some(self.position() + read_bytes as u64);
+    async fn read_vec(&mut self, ranges: &[Range<u64>]) -> io::Result<Vec<Bytes>> {
+        let absolute_ranges: Vec<Range<u64>> = ranges
+            .iter()
+            .map(|range| {
+                let start = self.offset + range.start;
+                let end = (self.offset + range.end).min(self.offset + self.size);
+                start..end
+            })
+            .collect();
+
+        for range in &absolute_ranges {
+            if range.start >= self.offset + self.size {
```

Comment:
Question: why not check this in line 72?

Reply:
udpated.